### ISSUE TITLE
Allowing to restart a batch data frame transform job will cause job to stay in "started" state

### ISSUE THREAD

**Issue by wwang500** [elastic/elasticsearch#46453]
Steps to reproduce:
- Create a data frame transform job `PUT _data_frame/transforms/...`
- Start it `POST _data_frame/transforms/.../_start`
- Once job finishes, try start again, `POST _data_frame/transforms/.../_start`

Observed:
 Job starts successfully and stay in "started" state for ever.

Expected:
a nice error message should be thrown, until we decide to support this case.

**Comment by sophiec20** [comment]
Please clarify what happens to the data? Are any changes made to the dest index? Are any searches performed on the source index?

**Comment by wwang500** [comment]
There is nothing happening on data. No searches were performed on the source index either.

**Comment by prwhelan** [comment]
Can repro, using [this example](https://www.elastic.co/guide/en/elasticsearch/reference/current/ecommerce-transforms.html) from our tutorials:
1. Load ecommerce sample data
2. Put transform:
```bash
PUT /_transform/ecommerce-customer-transform?pretty
{
  "source": {
    "index": [
      "kibana_sample_data_ecommerce"
    ],
    "query": {
      "bool": {
        "filter": {
          "term": {
            "currency": "EUR"
          }
        }
      }
    }
  },
  "pivot": {
    "group_by": {
      "customer_id": {
        "terms": {
          "field": "customer_id"
        }
      }
    },
    "aggregations": {
      "total_quantity.sum": {
        "sum": {
          "field": "total_quantity"
        }
      },
      "taxless_total_price.sum": {
        "sum": {
          "field": "taxless_total_price"
        }
      },
      "total_quantity.max": {
        "max": {
          "field": "total_quantity"
        }
      },
      "order_id.cardinality": {
        "cardinality": {
          "field": "order_id"
        }
      }
    }
  },
  "dest": {
    "index": "ecommerce-customers"
  },
  "retention_policy": {
    "time": {
      "field": "order_date",
      "max_age": "60d"
    }
  }
}
```
3. Start transform: `POST /_transform/ecommerce-customer-transform/_start`
4. Get transform should show `stopped` (keep clicking until it does): `GET /_transform/ecommerce-customer-transform/_stats`
5. Start transform: `POST /_transform/ecommerce-customer-transform/_start`
6. Get transform now always shows `started`: `GET /_transform/ecommerce-customer-transform/_stats`

The transform task just stays in the queue indefinitely: wakes up, sees there is no works, goes back in the queue, repeats.

**Comment by prwhelan** [comment]
Some observations:
- it looks like we can successfully `_stop` it, so that's good.
- letting the transform run and naturally finish might be an acceptable feature, but I don't know who is asking for it, and I don't presume to understand the implications of this kind of change.  There are a lot of places where we assume the task is either continuous or before the first checkpoint, so changing to allow a non-continuous transform to execute beyond the first checkpoint may have impact that I don't have the expertise to predict.
- there's at least some way to "rerun" finished batch transforms, via [reset](https://www.elastic.co/guide/en/elasticsearch/reference/current/reset-transform.html)

**Comment by prwhelan** [comment]
First working iteration:

- `POST /_transform/ecommerce-customer-transform/_start`
- `GET /_transform/ecommerce-customer-transform/_stats`
```bash
{
  "count": 1,
  "transforms": [
    {
      "id": "ecommerce-customer-transform",
      "state": "stopped",
      "stats": {
        "pages_processed": 2,
        "documents_processed": 4675,
        "documents_indexed": 46,
        "documents_deleted": 0,
        "trigger_count": 1,
        "index_time_in_ms": 7,
        "index_total": 1,
        "index_failures": 0,
        "search_time_in_ms": 22,
        "search_total": 2,
        "search_failures": 0,
        "processing_time_in_ms": 20,
        "processing_total": 2,
        "delete_time_in_ms": 5,
        "exponential_avg_checkpoint_duration_ms": 96,
        "exponential_avg_documents_indexed": 46,
        "exponential_avg_documents_processed": 4675
      },
      "checkpointing": {
        "last": {
          "checkpoint": 1,
          "timestamp_millis": 1707772244970
        },
        "changes_last_detected_at": 1707772244970
      },
      "health": {
        "status": "green"
      }
    }
  ]
}
```
- `POST /_transform/ecommerce-customer-transform/_start`
```bash
{
  "error": {
    "root_cause": [
      {
        "type": "status_exception",
        "reason": "Cannot start a non-continuous transform [ecommerce-customer-transform] that has already finished."
      }
    ],
    "type": "status_exception",
    "reason": "Cannot start a non-continuous transform [ecommerce-customer-transform] that has already finished."
  },
  "status": 409
}
```

--------------------------------------------

For the "let it run the next checkpoint," we may be able to just treat it like the next iteration in a continuous transform by changing these lines:
1. Remove the if/else clause in [TransformTask.triggered](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java#L460-L465) and always run `maybeTriggerAsyncJob`
2. Maybe also change [TransformIndexer](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java#L706) to always stop for non-continuous transforms regardless of whether changes were detected.

**Comment by prwhelan** [comment]
Current status:
*Most* of our integration tests will call `start` multiple times on a batch transform, which will reproduce this issue.  We'll need to refactor them to only check that the first iteration had been started before we can move forward with a change.

Though this issue is relatively benign, this fits the theme of "make sure Transforms reflects the correct state to the user" so we should continue with it. I'll continue working on it at a lower priority, since it can be stopped and it doesn't waste *too many* system resources to remain stuck in started (it doesn't actually run the search).

**Comment by prwhelan** [comment]
Related https://github.com/elastic/elasticsearch/issues/98802
